### ISSUE TITLE
Cannot Open Extension Control File

### ISSUE THREAD

**Issue by Smtgr14** [biolab/orange3#2083]
##### Orange version

V 3.4.0 (git revision 4438eb2)

##### Expected behavior
Adding and entering connection information to SQL Table widget shoul move to the next step.

##### Actual behavior
Error is thrown: 
could not open extension control file "/usr/share/postgresql/9.6/etx...

##### Steps to reproduce the behavior
1) Install a fresh copy of Orange (and python) on Windows 7 or Mac OS Sierra.
2) Run Orange3
3) Add SQL Table widget to canvas
4) Fill out all fields.
5) Error appears.

##### Additional info (worksheets, data, screenshots, ...)
<img width="424" alt="err" src="https://cloud.githubusercontent.com/assets/799384/23731789/17adbdce-0435-11e7-9ecd-ff370a908f72.png">

**Comment by azrulhaifan** [comment]
@Smtgr14 i have mac os sierra too, and i have issue in sql table too. but my error is "Please install a backend to use this widget", even that i installed psycopg2 the error is still appear.

can you help me ? sorry for bad english

**Comment by lexykassan** [comment]
Similar issue here. Orange 3.4.1 installed via Anaconda 1.6.3

psycopg2 is installed in the environment and seems to be running correctly. When using the SQL Table widget and entering credentials for a Google Cloud hosted PostgreSQL server I get the following error:

could not open extension control file "/opt/bitnami/postgresql/share/extension/quantile.control": no such file or directory 

Not even sure what this is meant to be or is looking for. Any help would be appreciated.

**Comment by ajdapretnar** [comment]
Possible duplicate of #2271. Could someone confirm it's the quantile issue?

**Comment by dirtyworks** [comment]
There is no way to know if it is a quantile issue on Windows as the error message does not fit in the GUI.

It seems like the default connection assumes  that the PostgreSQL database is local?  That simply isn't possible in my environment for many reasons. 
![2017-09-22 10_54_22-sql table](https://user-images.githubusercontent.com/3038055/30757954-5d73094a-9f85-11e7-95f8-cfc51fadc075.png)

**Comment by Smtgr14** [comment]
+1

**Comment by apetrov** [comment]
same on Orange3.7 + OSX 10.12.3

**Comment by Quazer** [comment]
same on Orange3.4.1 (Anaconda3-5.0.1-Windows-x86_64), PostgreSQL 9.6.6

**Comment by vpejovic** [comment]
Same issue with version 3.9.0.dev0 (slightly different error though - could not open extension control file "/usr/share/postgresql/9.4/extension/t...[can't see the rest]")

**Comment by ivek1312** [comment]
On gentoo linux and version 3.9.0.dev0 i get:
`could not open extension control file "/usr/share/postgresql/9.4/extension/tsm_system_time.control":No such file or directory`

Postgresql version is 10.1 installed in /usr/share/postgresql. 
Psycopg is 2.7.3.1.

**Comment by shahrzadpeyman** [comment]
Hi everybody!
I had the same problem:
```
could not open extension control file "/usr/local/share/postgresql/extension/quantile.control": no such file or directory
```
on:
```bash
# MacOS
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.13.2
BuildVersion:	17C88

# PostgreSQL 10(Brewed)
$ psql --version
psql (PostgreSQL) 10.1

# Python3(Brewed)
$ python3 --version
Python 3.6.4

# Orange3(Using Pip3 of Brew)
$ pip3 --version
pip 9.0.1 from /usr/local/lib/python3.6/site-packages (python 3.6)

$ pip3 show orange3
Name: Orange3
Version: 3.8.0
Summary: Orange, a component-based data mining framework.
Home-page: http://orange.biolab.si/
Author: Bioinformatics Laboratory, FRI UL
Author-email: info@biolab.si
License: GPLv3+
Location: /usr/local/lib/python3.6/site-packages
Requires: scikit-learn, setuptools, bottleneck, keyrings.alt, pip, dill, chardet, serverfiles, scipy, CommonMark, joblib, keyring, xlrd, pyqtgraph, AnyQt, numpy, docutils

$ pip3 show psycopg2
Name: psycopg2
Version: 2.7.3.2
Summary: psycopg2 - Python-PostgreSQL Database Adapter
Home-page: http://initd.org/psycopg/
Author: Federico Di Gregorio
Author-email: fog@initd.org
License: LGPL with exceptions or ZPL
Location: /usr/local/lib/python3.6/site-packages
Requires:
```

## How to solve it:

### Install [Quantile](https://github.com/tvondra/quantile) using [PGXN](https://github.com/pgxn/pgxn):
```bash
sudo easy_install pgxnclient
pgxn install quantile
```
This will install the missing `quantile.control` at `/usr/local/share/postgresql/extension/quantile.control`.
### Run Orange
Just run orange now:
```bash
orange-canvas
```
Finally Add SQL Table widget to canvas and Enjoy

**Comment by marcot21** [comment]
> Install Quantile using PGXN:
> sudo easy_install pgxnclient
> pgxn install quantile

Trying to install Quantile to solve issue:  
could not open extension control file "/usr/share/postgresql/9.6/etx...
<img width="1280" alt="schermata 2018-03-04 alle 20 47 36" src="https://user-images.githubusercontent.com/37043719/36949788-6527870e-1fed-11e8-9905-b7eb77614942.png">

Using MAC OS High Sierra 10.13.3. 
Postgre 9.6 (Pgadmin)
Orange 3.1 (anaconda)

When launch:
*pgxn install quantile*

got following error:
'''INFO: best version: quantile 1.1.5
INFO: saving /tmp/tmp7qgsiw6v/quantile-1.1.5.zip
INFO: unpacking: /tmp/tmp7qgsiw6v/quantile-1.1.5.zip
INFO: building extension
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk -mmacosx-version-min=10.7 -arch i386 -arch x86_64 -O2   -I. -I./ -I/Library/PostgreSQL/9.6/include/postgresql/server -I/Library/PostgreSQL/9.6/include/postgresql/internal -I/opt/local/20151229/include/libxml2  -I/opt/local/Current/include/libxml2 -I/opt/local/Current/include  -c -o src/quantile.o src/quantile.c
clang: warning: no such sysroot directory: '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk' [-Wmissing-sysroot]
clang: warning: no such sysroot directory: '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk' [-Wmissing-sysroot]
src/quantile.c:7:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
make: *** [src/quantile.o] Error 1
ERROR: command returned 2: make PG_CONFIG=/Library/PostgreSQL/9.6/bin/pg_config all''

**Comment by Kegerer** [comment]
Hi All,

  I'm new to this, but am having the same/similar issues as above.

Orange version: 3.13.0
Anaconda 3
PostgreSQL 10 - pgAdmin4
Add-on: pyscopg2 successfully added

**Expected behavior:**
After correctly adding pyscopg2 in the Orange Add-ons, connection to pgSQL database should be possible.

**Actual behavior:**

**Scenario 1:** Fill in the SQL Table connection information completely produces the known and previously described "extension control file" error message. For me it looks like this: 
![grafik](https://user-images.githubusercontent.com/39377557/40230031-753ac5dc-5a96-11e8-9313-d04f41bacb93.png)

**Scenario 2:** Fill in the SQL Table connection information all fields except "username", and the connection to the pgSQL database seems to work. The only problem is, when connecting the SQL Table Widget to the Data Table Widget, the line is dotted (indicating that the connection or reading of data from the SQL table is not working (?)). See here: 
![grafik](https://user-images.githubusercontent.com/39377557/40230192-f5cc8e4c-5a96-11e8-8cbd-d9939ef6c3c5.png)

**Steps to reproduce behavior:**

   1. Install a fresh copy of Orange 
   2. Run Orange3
   3. Add SQL Table widget to canvas having loaded the Add-on pyscopg2 
   4. Fill out all fields.
   5. Error appears (paritally in SQL table: see first image above)



**My thoughts as an inexperienced sql user:**
Scenario 2 seems some how wrong because one would expect being able to select a table from the pgSQL database server (similar to how one ones a file on the orange canvas), but this is not possible.

Scenario 1: Following shahrzadpeyman's comment: 

> How to solve it:
Install Quantile using PGXN:

```
sudo easy_install pgxnclient
pgxn install quantile
```

> This will install the missing quantile.control at /usr/local/share/postgresql/extension/quantile.control.
> Run Orange
> 
> Just run orange now:
> 
> orange-canvas
> 
> Finally Add SQL Table widget to canvas and Enjoy 

... I think I need to install Quantile using PGXN. At this point, I am lost due to my inexperience. Looking at pg's description of what a PGXN is [https://wiki.postgresql.org/wiki/PGXN](url) , I'm not certain if I have this and if not, how I activate or install it. Furthermore, looking at [https://github.com/tvondra/quantile](url), how do I know I am using the pgxn client? 
 
 Where/how do I complete the commands: 
`sudo easy_install pgxnclient`
`pgxn install quantile`  ?

Or maybe I'm just missing something. Your help is appreciated!

**Comment by AlexandreVR** [comment]
Hello all,

I am having the same problem. I already installed psycopg2 but now I cant open the sql table in Orange, because it presents the same error: "Could not open extension control file (...)quantile.control"

I am using a windows 10 machine, and it is being very hard to install quantile. Right now I already tried to install pgxn client, in the script folder inside orange folder, using pip install pgxnclient, but then when I do pgxnclient it presents: "File "c:\users\maria sousa\appdata\local\orange\lib\site-packages\pgxnclient\cli.py", line 114, in command_dispatch
    os.execv(argv[0], argv)".

I also tried to just install quantile, with pip install quantile, but the erros mantains.

I already tried lots of things to try to install quantile. The thing is, I really need to use this tool for my work. 

Your help would be very much appreciated!

Thank you very much.

**Comment by Witenite** [comment]
Hi,
same issue here, except I am running on Windows (only because I have to as that's what we use at work...) I can confirm that this is the same problem as the missing quantile control, except mine keeps telling me I'm missing the tsm_system_time.control file (visible when hovering the mouse over the error). There just seems to be no way to install this extension, unless you are running Linux.

So what's the answer? How do you install these missing controls/extensions under a Windows 7 (client, with Windows 2012 Server running Postgres database)? We are using Postgres version 9.2.9 which, while it's old, seems to be no worse off than the latest Postgres version 10 when it comes to this problem.

Thanks!

**Comment by AlexandreVR** [comment]
Astaric thank you, but  that link you posted does say that you need to install quantile but it does not gives instructions about how to do it, and for me, that is the real problem.

**Comment by astaric** [comment]
The wiki page linked to the page with instructions for installation of the quantile extension, but they were buried at the bottom of the site. I have added the commands needed to install the extension to the [wiki page](https://github.com/biolab/orange3/wiki/Installation-of-PostgreSQL-extensions).

There is also an open pull request - https://github.com/biolab/orange3/pull/3113 - which should allow you to use the database even without extensions. If anyone is able to try it out, let us know what works/does not work in PR comments.

**Comment by AlexandreVR** [comment]
Thank you for your answers astaric. However I am still having some problems. 

I could not install quantile using pgxnclient, because it is presenting the following error when I use pgxn:
![image](https://user-images.githubusercontent.com/24398171/42481130-bc83ed0e-83d8-11e8-9463-979175d38d7b.png)

Therefore, I was trying to use the make install, but I dont know where to place (in what folder) the downloaded file from the quantile website, in order to use the make install command.

I didn't tried yet to use the database without extensions because I wanted to try these two steps first.

Thank you very much.
